Add empty state help for projects page

spadgett opened the pull request:
Adds some help text when you first login to OpenShift and have no projects.

spadgett commented:
@jwforres @smarterclayton @liggitt @ajacobs21e what do you think?

Here is an empty project for reference. We might change that message because we're repeating ourselves.

<img width="790" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/9415956/ba3502e2-4810-11e5-82fb-0e0e4e2373e0.png">

spadgett commented:
I wanted to add an OpenShift logo beside the "Welcome to OpenShift" header, but I couldn't find one that worked on a gray background.

jwforres commented:
Couple comments:
* in v3 we dont refer to them as QuickStarts, we have been calling them instant apps
* don't say "community" documentation since this will be in enterprise as well and enterprise would link to the enterprise docs
* I think we should still have the same type of help when you can't create projects, just change the "create a project" references as needed to say request a project or something

@sg00dwin see sam's comment about wanting an openshift logo https://github.com/openshift/origin/pull/4317#issuecomment-133524532

spadgett commented:
We should probably open a doc defect since it uses the term QuickStart:

https://docs.openshift.org/latest/admin_guide/install/first_steps.html#creating-quickstart-templates

liggitt commented:
careful with logos or community references... those all need to be swappable in alternate releases

liggitt commented on assets/app/views/projects.html:
```diff
@@ -1,28 +1,94 @@
+    </div>
+
+    <div ng-if="!canCreate" class="empty-state-msg">
+      <p class="lead">Unfortunately, you don't have authority to create projects.</p>
```
remember, `newProjectMessage` can contain the customized message from the admin. They might have enabled a way for users to create projects for themselves, so we have to show it if it exists (using `linky`), and we shouldn't pair it with an "Unfortunately..." header

jwforres commented:
Ok I wasn't aware of that, I guess we started calling them QuickStarts
again? @smarterclayton ? @bparees?

On Fri, Aug 21, 2015 at 2:42 PM, Sam Padgett <notifications@github.com>
wrote:

> We should probably open a doc defect since it uses the term QuickStart:
>
>
> https://docs.openshift.org/latest/admin_guide/install/first_steps.html#creating-quickstart-templates
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-133527441>.
>

bparees commented:
@jwforres i'm not sure we ever did a great job of picking one or the other.

v2 quickstart:  a way to get multiple cartridges created at once.
v2 instant-app: a way to get an app created from a specific source repo.  may also include multiple cartridges, so arguably an enhanced quickstart

v3 template: can serve in the role of either a quickstart or an instant-app (might provide a default source-uri to a useful app, or might expect you to provide a source-uri because it's just a generic helper for building java apps with a database)

but i'm ok with changing the docs to s/QuickStart/InstantApp/ if that's what we want.

smarterclayton commented:
I'm not in love with the white box.  I would prefer something more subtle.  To many overlapping rectangles.

jwforres commented:
which white box... the one on the empty project list is exactly what we do
in v2.  Are you talking about the one after you are in a project and the
project is empty?

On Fri, Aug 21, 2015 at 3:03 PM, Clayton Coleman <notifications@github.com>
wrote:

> I'm not in love with the white box. I would prefer something more subtle.
> To many overlapping rectangles.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-133532377>.
>

smarterclayton commented:
Yes, project is empty.  Basically when it's white left nav and white top bar, then gray, then white box, it looks out of balance.  I think empty should look empty.

liggitt commented:
I like it. Empty grey feels sad. Sad is bad.

smarterclayton commented:
In that first image though, what does it look like when you don't have the white box on the empty projects page?  Flush left alignment with no box border.

smarterclayton commented:
Or instead of the centered box, do the full width page white bar (like what we have on the homepage and other pages).

spadgett commented:
Updated for the comments from @jwforres and @liggitt. Using Instant App for now, but I can change it back.

@smarterclayton working on some screens of your suggestions

spadgett commented:
@smarterclayton Gray background. Maybe too many different fonts.

<img width="799" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/9417210/4712740e-4818-11e5-9f7d-d67e347301e1.png">

liggitt commented:
the big link feels unbalanced... it looks as big as the header. Should it be in the "Create a project" portion of the outline?

smarterclayton commented:
That's a lot of fonts.

spadgett commented:
White edge to edge background.

<img width="885" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/9417508/f7964768-481a-11e5-91f7-29e34b6625c5.png">

I can remove some of the fonts.

spadgett commented:
Fewer fonts:

<img width="886" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/9417557/569c7d0e-481b-11e5-9d0e-ed4254507e26.png">

smarterclayton commented:
The center content would look better with two columns - sentences that wrap the page are unreadable for most humans.  This really needs to be a flow from 1 to 2a and 2b - I would organize them visually that way.

smarterclayton commented:
Or even make the right column for the center be a picture so it's narrower text.  And cut out some of the text.

spadgett commented:
Yeah, typically we add generous margins for wide screen widths, although my hack to try edge to edge white removed them. I think the columns are a good ideas, but I'd probably need to stack them on mobile.

smarterclayton commented:
Stack for sure - a picture would be better even.

On Fri, Aug 21, 2015 at 3:55 PM, Sam Padgett <notifications@github.com>
wrote:

> Yeah, typically we add generous margins for wide screen widths, although
> my hack to try edge to edge white removed them. I think the columns are a
> good ideas, but I'd probably need to stack them on mobile.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-133544862>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

spadgett commented:
@smarterclayton I wouldn't use this image, but is this what you had in mind?

<img width="773" alt="openshift_web_console" src="https://cloud.githubusercontent.com/assets/1167259/9418259/aa1da282-4820-11e5-9e82-b939eb8770a5.png">

smarterclayton commented:
Yes

smarterclayton commented:
Bonus points for having a graphic with arrows that shows source code going into a box and then the box being sent out on a conveyor line (or carried by helicopters into a cloud)

jwforres commented:
cc @sg00dwin

spadgett commented:
The problem with the edge-to-edge white is the long lines... It looks strange with wide margins (see screenshot). I think it'll still be a problem even with an image unless it's very large.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9441368/b2b52454-4a43-11e5-8e24-b9f99c6f0a31.png)

ajacobs21e commented:
I have two suggestions:
![empty state](https://cloud.githubusercontent.com/assets/6415742/9441603/1af3fd8c-4a45-11e5-92a8-78ddb44dcac5.png)
With minor changes to the messaging - whatever messaging we use should be reviewed by a content writer.

The info is good but it may be more useful if split up onto the relevant pages:
![empty state2](https://cloud.githubusercontent.com/assets/6415742/9441756/ec09e1ac-4a45-11e5-9282-9ecd47b420e5.png)

![empty state3](https://cloud.githubusercontent.com/assets/6415742/9441832/53558e38-4a46-11e5-9448-b255e5161d9e.png)


The white stripe on the background looks strange and doesn't fit with the OpenShift or Patternfly theming.

liggitt commented:
+1 for splitting the content into multiple pages to just show the next relevant step

spadgett commented:
OK, starting to split up the content.

I'm not sold on gray without the box. It feels too empty. Maybe if vertically centered, but honestly the gray is kind of dull.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9442454/049f1274-4a4a-11e5-9c62-9812705613bf.png)

Haven't really changed the styles here (yet), but changed the wording.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9442672/138b28da-4a4b-11e5-9590-88f44ad057ae.png)

We tried to be careful with the message before because the project might not really be new or empty, even if we didn't show anything in the overview.

ajacobs21e commented:
The empty state is supposed to look empty to make you want to click that button and fill it in with pretty stuff. 

I'm good with taking out "new" project. Is there a content writer we work with for OpenShift?

smarterclayton commented:
I'm a little concerned with the white box because we use it to mean "this is a thing in the system you interact with".  The empty state is not quite the same thing.  It sounds more to me like we're lacking the visual language to identify a "take me places, get me started".  And centering it without the box leaves it a bit lost and alone on the page doesn't help strengthen that message.  

Why use small font styles?  If you have 100 or so words to use, and they are important, why not put them into two or three lines together at an H2 level font size?  The suggestion of images was really to create and fill space that you can balance the text against.

All that being said, anything is better than nothing right now, so I'm fine with doing something to get it moving.  We can always fine tune later or have someone spend some time crafting a more impressive visual presence.

spadgett commented:
Project page without the white box for comparison. This one doesn't feel so empty to me. I can play with font sizes and spacing for the empty project list.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9444121/d1ef5718-4a52-11e5-80f3-0b3cb8216862.png)

sg00dwin commented:
Recommend removing the h1 heading with line, to pull things up.

smarterclayton commented:
Left margin is pretty ragged - that's a case where a left image might serve to give separation from the left nav without requiring extra constructs.  Either way, not a huge deal.

spadgett commented:
More top margin on the all projects page:

![screen shot 2015-08-24 at 1 21 20 pm](https://cloud.githubusercontent.com/assets/1167259/9447014/0f15428c-4a63-11e5-97d3-c85a5c836f2b.png)

Mobile:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9447071/4bdc0bce-4a63-11e5-8157-2daa43de821f.png)

jwforres commented:
I just want to make sure we don't lose the requirement that we explain what a project is, that was the complaint we had.

sg00dwin commented:
If the project is empty, then it would seem we shouldn't show the left bar primary nav items.

spadgett commented:
@sg00dwin There might actually be content there, though. For instance, we filter some pods on the overview, but they might appear in the browse pods page. That's why I want to be a little careful with the language.

Settings might have content like quotas, too.

jwforres commented:
I disagree from the perspective that a project admin might create a
project, go to settings / members to give a bunch of people access, but
never want to add anything to the project themselves.
On Aug 24, 2015 1:34 PM, "Steve Goodwin" <notifications@github.com> wrote:

> If the project is empty, then it would seem we shouldn't show the left bar
> primary nav items.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-134310211>.
>

spadgett commented:
Removed `h1` and left-aligned content.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9447818/4fce9284-4a67-11e5-859c-924ca270a7c5.png)

spadgett commented:
@jwforres opinion on the icons?

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9469124/4a76473a-4b13-11e5-8e7f-df9fe2e5f117.png)

spadgett commented:
And w/o the negative margin:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9469200/c2acbd38-4b13-11e5-85a7-c6e53692d073.png)

spadgett commented:
OK last one:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9469285/390498f2-4b14-11e5-9f48-519abc2ca9df.png)

spadgett commented:
Unrelated travis failure:

```
--- FAIL: TestRequestExecuteRemoteCommand-2 (0.11s)
	remotecommand_test.go:112: 2: stdin: expected "a", got ""
FAIL
```

ajacobs21e commented:
@spadgett out of the three options I like #1 the best (icons past left margin) however I don't think the icons add anything - unless we also use these same icons on the create flow.

jwforres commented:
LGTM, we'll follow up with a picture for the initial welcome page when we actually have one to use

last chance for any arguments

jwforres commented:
@ajacobs21e we use the thunderbolt on templates that don't specify their own icon

ajacobs21e commented:
@jwforres ok cool. what do you think about using the icons on the initial Create page with the 'From Source' and 'From Template; headings?

jwforres commented:
@ajacobs21e makes sense to me, would like to see how it looks with them there

spadgett commented:
![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9474817/d9942758-4b30-11e5-99b0-8796252a5a4a.png)

spadgett commented:
@jwforres Pushed the icon changes to create.html and fixed the icon widths. PTAL

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/9475578/8314507e-4b35-11e5-9b05-7778115f0cd4.png)

jwforres commented:
[merge]

smarterclayton commented:
That merge looked like it had some angst behind it :)

On Tue, Aug 25, 2015 at 3:37 PM, Jessica Forrester <notifications@github.com
> wrote:

> [merge]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-134696146>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

jwforres commented:
Test flake [merge]

liggitt commented:
for posterity:
```
1)  authenticated e2e-user when creating a new project should successfully create a new project
  - Expected false to be true.
      at /data/src/github.com/openshift/origin/assets/test/e2e/test.js:303:91
```

spadgett commented:
> Expected false to be true.

Words to meditate on.

liggitt commented:
Get used to disappointment

smarterclayton commented:
Expected false to be FILENOTFOUND

On Wed, Aug 26, 2015 at 11:59 AM, Jordan Liggitt <notifications@github.com>
wrote:

> Get used to disappointment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-135050505>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

jwforres commented:
@spadgett the test case failure is because you are hiding the project h1 now when the project is empty

https://github.com/openshift/origin/blob/master/assets/test/e2e/test.js#L303

spadgett commented:
@jwforres Thanks, I removed that line since the `h1` is no longer there. I'm not sure if `goToAddToProjectPage` will fail because we added the icons.

[test]

jwforres commented:
Hmm instead of removing it can you change it to check for the project name
in the project selector dropdown?

On Wed, Aug 26, 2015 at 12:58 PM, Sam Padgett <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> Thanks, I removed that line since
> the h1 is no longer there. I'm not sure if goToAddToProjectPage will fail
> because we added the icons.
>
> [test]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-135107690>.
>

spadgett commented:
@jwforres I can give it a shot, but the e2e tests don't run reliably for me in Vagrant. Is this the right command?

```
TEST_ASSETS=true TEST_ASSETS_HEADLESS=true hack/test-end-to-end.sh
```

 Is there a way to run just the UI tests?

jwforres commented:
On Wed, Aug 26, 2015 at 2:03 PM, Sam Padgett <notifications@github.com>
wrote:

> @jwforres <https://github.com/jwforres> I can give it a shot, but the e2e
> tests don't run reliably for me in Vagrant. Is this the right command?
>
> TEST_ASSETS=true TEST_ASSETS_HEADLESS=true hack/test-end-to-end.sh
>
> Yes.  Although I don't know if the latest vbox has been updated to include
chrome.  Have you tried launching an ec2 instance instead, syncing your
source and then running the tests? Its a bit slow rebuilding everything,
but should work.

>  Is there a way to run just the UI tests?
>

Unfortunately not today @liggitt started to look at whether it would be
possible.  Right now the other end to end tests are populating the empty
state of the project for the first test.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4317#issuecomment-135125549>.
>

spadgett commented:
[test]

spadgett commented:
@jwforres I updated the test. Please take a look.

jwforres commented:
[merge]
